File: Algorithm_130_Surrounded Regions.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    void solve(vector<vector<char>>& board)
    {
        int m=board.size();
        if (m==0) return;
        int n=board[0].size();
        vector<vector<int> > surrounded(m, vector<int>(n));
        vector<vector<int> > visited(m, vector<int>(n));
        for (int i=0;i<m;i++) for (int j=0;j<n;j++)
        {
            surrounded[i][j]=1;
            visited[i][j]=0;
        }
        for (int i=0;i<m;i++)
        {
            if (board[i][0]=='O' && !visited[i][0]) dfs(board,surrounded,visited,i,0);
            if (board[i][n-1]=='O' && !visited[i][n-1]) dfs(board,surrounded,visited,i,n-1);
        }
        for (int j=0;j<n;j++)
        {
            if (board[0][j]=='O' && !visited[0][j]) dfs(board,surrounded,visited,0,j);
            if (board[m-1][j]=='O' && !visited[m-1][j]) dfs(board,surrounded,visited,m-1,j);
        }
        for (int i=0;i<m;i++) for (int j=0;j<n;j++) if (board[i][j]=='O' && surrounded[i][j]) board[i][j]='X';
        return;
    }
    void dfs(vector<vector<char> > &board, vector<vector<int> > &surrounded, vector<vector<int> > &visited, int i, int j)
    {
        int m=board.size(), n=board[0].size();
        if (i<0 || i==m || j<0 || j==n || visited[i][j] || board[i][j]=='X') return;
        visited[i][j]=1;
        surrounded[i][j]=0;
        dfs(board,surrounded,visited,i-1,j);
        dfs(board,surrounded,visited,i+1,j);
        dfs(board,surrounded,visited,i,j-1);
        dfs(board,surrounded,visited,i,j+1);
    }
};

File: Algorithm_130_Surrounded Regions.java
// 思路：先从四条边界上做DFS，把所有与边界相连的'O'标记成'A'
// 然后遍历board，把所有的'A'变回'O'，把所有的'O'变为'X'

class Solution {
    public void solve(char[][] board) {
        if (board == null || board.length == 0 || board[0] == null || board[0].length == 0) return;
        for (int col = 0; col < board[0].length; col++) { // top border and bottom border
            mark(board, 0, col);
            mark(board, board.length - 1, col);
        }
        for (int row = 0; row < board.length; row++) { // left border and right border
            mark(board, row, 0);
            mark(board, row, board[0].length - 1);
        }
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (board[i][j] == 'A') board[i][j] = 'O';
                else board[i][j] = 'X';
            }
        }
    }
    private void mark(char[][] board, int i, int j) {
        if (i < 0 || i == board.length || j < 0 || j == board[0].length || board[i][j] != 'O' ) return;
        board[i][j] = 'A';
        mark(board, i - 1, j);
        mark(board, i + 1, j);
        mark(board, i, j - 1);
        mark(board, i, j + 1);
    }
}